src/semantics.js

```diff
@@ -0,0 +1,93 @@
+/* eslint-disable unicorn/no-null */
+
+import {
+  CapabilityEscalationError,
+  CapabilityParseError,
+  CapabilityUnrelatedError,
+} from './errors.js'
+
+const uploadLevels = {
+  'upload/*': 0,
+  'upload/IMPORT': -1,
+}
+
+const PREFIX = 'storage://'
+/**
+ * @type {import('./types').CapabilitySemantics<import('./types').StorageSemantics>}
+ */
+export const storageSemantics = {
+  tryParsing(cap) {
+    if (typeof cap.with !== 'string') {
```

Comment:
NTH. here and across the module would be quite nice to have a more handle data validation in a declarative way rather than imperative